### PR TITLE
Select, unselect, and split a spool along the fiber

### LINKED ISSUE #857: Selecting on inventory-defined fields is not supported when an inventory is attached
Once a spool carries an inventory, the fields that inventory defines should be selectable, and today they are not: `spool.attach_inventory(inv).select(coupling="cement")` raises `InvalidSpoolQueryError` because the index has no such column. The spec (`API/spool.qmd`) documents this workflow, so the gap is between the design and the implementation rather than in the design.

Two levels, with quite different mechanics:

**Acquisition-level (whole patch).** Fields of the acquisition or interrogator — `gauge_length`, `interrogator.model`. A patch matches or it does not. Resolution is pure metadata: for each distinct `data_source_id` and time window in the index, resolve the acquisition, test the predicate, keep or drop the row. No patch data is loaded and `len(spool)` stays exact.

**Channel-level (within a patch).** Track fields (`coupling.medium`), annotation groups (`zone`), geometry axes (`latitude`, `z`), and optical `distance`. These select a set of channels, which may be disjoint along the fiber, so a matching patch is trimmed and sometimes subdivided. That changes the number of patches, so it cannot be answered from the index alone; it needs a resolution pass over the inventory per row. `prune_to_inventory` (below) is the natural place for that pass, since it is already the documented one-time metadata cost.

**What comes back on the patch.** A name used in a select should arrive on the extracted patch, and nothing more: `select(coupling="cement")` yields patches carrying `coupling.medium` so the result says why each channel survived, without paying for blanket enrichment. Composed with an explicit `Spool.enrich`, the patch owes the union of what was asked for.

Related and not yet implemented:

- `Spool.prune_to_inventory(inventory=None)` — resolve the index against the inventory, drop rows it does not describe (`on_missing` of `raise`/`warn`/`drop`), and subdivide rows straddling an epoch boundary.
- Lazy attachment: `attach_inventory` should accept a path and not read it until something needs it, plus an auto-attach from a blessed name in a spool directory. This depends on the authoring-format loader, which does not exist yet.

`Spool.select` currently adds a note to its error naming the attached inventory and saying this is not supported yet, so the failure at least points somewhere useful.

### PR DESCRIPTION
**Pull request by d-chambers**
## Description

Phase 3 (c) of the inventory work (#857), the last of the three: selection along the fiber. Closes #857.

A track, an annotation group, or a geometry axis describes *channels* rather than whole patches, so selecting on one trims each patch to the channels which match, and subdivides it where the matching region is disjoint. `len` can grow, exactly as it can under `conform_to_inventory`.

```python
spool = dc.spool("/data/archive").attach_inventory(inventory)

raw = (
    spool.conform_to_inventory()
    .unselect(coupling=None)          # drop the channels with no coupling
    .select(zone="north")             # keep the ones in one annotation group
    .enrich()
)

by_zone = spool.split_by("zone")      # one patch per value, stamped with it
```

**None of the selection semantics are new, and that is the point.** Values are projected onto the channels by the same function `Patch.enrich` projects them with, and judged by the same predicate the index applies to a stated attr. So globs, sequences, ranges, and booleans mean here exactly what they mean there, and — the property worth having — every channel a selection keeps is one enrichment would give the value asked for. Only `None` needed a reading of its own: it is how a query spells the undefined marker `_fill_from_intervals` writes (`""`, NaN, or False, by kind).

**The complement is exact.** The mask runs along one dimension, so unlike a patch's rectangle, what a selection did not keep is expressible as channels. Naming acquisition-level facts alongside channel-level ones still yields *one* complement: a patch the attrs never matched keeps every channel, because the selection it complements never held it. Complementing the two halves apart would drop that patch.

**`Patch.unselect`** lands here because channel trimming consumes it (settled on #857). It asks `select` itself which samples it would keep and inverts the answer, so a selector cannot come to mean one thing in `select` and another in its complement. Naming several coordinates complements each on its own — the true complement of a block is an L, which no array can hold.

**`Spool.split_by`** is the same subdivision asked differently: each value a group takes decides an output. Every kind splits, and absence is not a value — except in a membership group, where `False` says something about every channel rather than the absence of one.

### Design notes

*Subdivision generalized from cuts to pieces.* `conform_to_inventory` partitions a span at cut values; a channel query keeps runs and drops the gaps, so a row it matches nothing of leaves the spool and one it matches entirely passes through untouched. Cuts cannot say either, so `build_subdivision_plan` now takes the pieces and `subdivision_pieces` turns cuts into pieces beside it. A piece is "modified" when it differs from its own row rather than when its row had cuts — the same answer for a cut row, the right one for a kept one.

*Composition is by nesting, not a two-dimensional plan.* A distance plan on top of conform's time plan is lossless and exact; verified by reassembling the four pieces byte-identically. Selecting twice nests rather than collapsing, since a second selection must narrow rather than re-plan from source — the opposite of what re-chunking one dimension wants.

*Only a real dimension is trimmed.* What a non-dimensional coordinate says about the dimension it runs along is not in the index, so a patch which can only be placed by one raises rather than being answered for differently than enrichment would. A name the index already uses keeps its own meaning outright: `select(distance=...)` is the patch's own axis whether or not the inventory could also place it on the fiber. That last one was a real bug the tests caught — `_coords={"distance": ...}` was being read as optical distance.

*Grid division, again.* #882 replaced `to_float` on both operands with native division, and the plan predicted the two correction loops would become unreachable. That is true for a time axis (datetime64 is integer nanoseconds) and false for a float one, which is what subdividing along `distance` introduced. Measured over 400k random float triples: the ratio lands a hair high on ~5% of on-grid cuts — `1.0 + 0.1` differenced back out is `0.10000000000000009` — and a hair low on ~6% of off-grid ones. Both loops are load-bearing along the fiber, and both now have ordinary tests rather than a coverage pragma.

### Review notes

The six-reviewer pipeline (Codex plus five Claude reviewers) ran before this PR and found 34 items across the six legs. Four are worth calling out, since all four would have shipped:

- **A lossy plan must never collapse.** Re-planning the same dimension collapses onto a plan's members, which is sound only because a chunk's or a subdivision's members together *cover* their sources. Channel selection is the first plan here whose pieces do not, so `select(coupling=...).chunk(distance=...)` loaded back every channel it had removed while `get_contents` went on describing the ones it kept. A plan now records whether it drops samples — read off the pieces rather than taken on trust, since a plan wrongly called lossless resurrects data while one wrongly called lossy only forgoes a collapse.
- **Three reviewers independently found the same one-line omission**: the channel grid was rebuilt with the *signed* step, so a reverse-sorted patch counted to a negative number of samples and was silently dropped by every fiber query. `subdivision_pieces` has the guard and a comment saying why; the newer function did not.
- **Prose and test-vacuity independently found that `split_by` promised something impossible** — that a channel could land in more than one output of a single call. It cannot: a channel resolves to one value of a group, the same one enrichment projects. The test named for the sharing asserted the partition it actually is. The claim came from the spec, which is corrected too.
- **Reviewers that ran the code found two `Patch.unselect` failures** no reader would have predicted: two coordinates on one dimension, and a dimension carrying no values of its own. Both are fixed by combining the complements per dimension and applying them as sample numbers.

The review also found five tests that asserted only inside a loop over a spool whose length they never pinned — so an implementation selecting *nothing* passed all five, including the one test that selection and enrichment agree — and that nothing was red without the grid arithmetic #882 changed. That last is now a property over 900 random float grids, verified red with the correction loops removed.

## Changelog

- added: `Spool.select` and `Spool.unselect` on the coordinates an attached DASDAE inventory defines along the fiber — typed tracks, annotation groups, and geometry axes. These trim each patch to the matching channels and subdivide it where the matching region is disjoint, so a selection can change both the shape and the number of patches.
- added: `Spool.split_by`, which expands a spool into one patch per value of an inventory-derived coordinate, stamping each output with the value it was split on.
- added: `Patch.unselect`, the complement of `Patch.select`: it takes the same selectors and removes exactly the samples that selection would have kept.

## Checklist

I have:

- [x] filled in the Changelog section above (see `docs/contributing/general_guidelines.qmd`).

I have (if applicable):

- [x] referenced the GitHub issue this PR closes.
- [x] documented the new feature with docstrings and/or appropriate doc page.
- [x] included tests. See [testing guidelines](https://dascore.org/contributing/testing.html).
- [ ] added the "ready_for_review" tag once the PR is ready to be reviewed.

## Summary by CodeRabbit

* **New Features**
  * Added `unselect` to remove samples matching coordinate, channel, relative, or sample-based criteria.
  * Added inventory-aware channel selection, unselection, and splitting.
  * Added `split_by` for grouping spool data by inventory values with include/exclude filters.
  * Improved handling of uneven sampling, disjoint channel selections, and incomplete inventory coverage.
  * Improved unit-aware attribute filtering by converting values before comparison.
* **Documentation**
  * Added tutorial guidance and examples for complementary selection with `unselect`.
* **Tests**
  * Expanded coverage for coordinate complements, inventory operations, subdivision, units, and edge cases.

### CODE REVIEW

File: dascore/core/spool.py [hunk 1]
```diff
@@ -1396,16 +1714,50 @@ def conform_to_inventory(
...
+        onto the sources — which would load back the samples the pieces
+        left out.
+        """
+        from dascore.io.index.planned import derived_catalog  # noqa: PLC0415
+
+        plan = build_subdivision_plan(rows, pieces, name)
+        stamped = ()
+        if stamp is not None:
+            stamp_name, values = stamp
+            plan = replace(plan, outputs=plan.outputs.assign(**{stamp_name: values}))
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/core/spool.py:1781]: Fixed in fa28959e — and confirmed it really does corrupt the catalog: a group named `output_id` gave `ValueError: invalid literal for int() with base 10: 'x'` out of `derived_catalog`. `RESERVED_GROUP_NAMES` guards the names a group would shadow as a *coordinate*, so it says nothing about the plan's own bookkeeping. `split_by` now refuses to stamp over `output_id`, `dims`, an envelope column, or any private one; `stamp=False` still splits on such a group, since only recording the value is the problem. Test: `TestChannelReviewFindings::test_a_stamp_cannot_overwrite_the_plan`.

File: dascore/core/spool.py [hunk 2]
```diff
@@ -1019,24 +1163,38 @@ def _split_inventory_query(self, _attrs, _coords, kwargs, samples):
...
+            requested, names, known_attrs, known_coords, _coords, kwargs
         )
+        # Neither keyword has anything to mean about a value the fiber
+        # states: it has no sample numbering of its own -- the channels it
+        # describes are the patch's -- and no endpoints to be relative to,
+        # since what it says varies from one acquisition to the next.
+        # Ignoring either quietly would answer a question nobody asked.
+        for flag, label in ((samples, "samples"), (relative, "relative")):
+            if not (channels and flag):
+                continue
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/core/spool.py:1209]: Fixed in fa28959e. The flags are now judged against `_stated_channels(channels)` rather than every recognized name, so `select(distance=(0, 10), coupling=..., samples=True)` works and a selector which actually asks something still refuses the flag. Test: `test_a_no_op_channel_selector_does_not_veto_a_flag`.

File: dascore/proc/inventory.py [hunk 3]
```diff
@@ -770,6 +784,345 @@ def _get_coord_values(inventory, path, name, distances):
...
+        # A name this path defines nowhere states nothing about any of its
+        # channels, so it matches none of them -- listing a name is not
+        # promising a value for it.
+        return np.zeros(len(distances), dtype=bool)
+    if selector is None:
+        # `None` is the query spelling of the undefined marker, which is a
+        # value here rather than the "select everything" a bare None means
+        # of an attr: a channel the track says nothing about is a channel.
+        return _undefined_mask(values)
+    return evaluate_attr_predicate(list(values), name, selector, units)
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/proc/inventory.py:1123]: Fixed in fa28959e, though not quite where you pointed. The conversion was not missing at the call site — `evaluate_attr_predicate`'s scalar path validated it and then discarded the converted value, where `_range_bounds` keeps its bounds. Reproduced directly: values `[10, 20, 30]` in degrees against `20 * degree` gave `[False, False, False]` while the range form matched. `selector` now returns the converted value. It is in the index's own predicate and applies to a stated attr too, so it was reachable before this branch — but routing coordinate values through it makes it much more so. Test: `test_a_scalar_quantity_selector_is_converted`.

File: dascore/core/spool.py [hunk 4]
```diff
@@ -1297,6 +1512,105 @@ def enrich(
...
+            self._inventory, contexts, working, name, _glob_filter(include, exclude)
+        )
+        _refuse_rows(source_rows, reasons, _UNPLACEABLE)
+        if dim is None:  # nothing to split: no row has a fiber to split on
+            return self._restrict_to_rows([])
+        pieces = [[piece for _, piece in row] for row in rows]
+        marks = None
+        if stamp:
+            marks = (name, [value for row in rows for value, _ in row])
+        return self._subdivided(source_rows, working, pieces, dim, marks)
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/core/spool.py:1636]: Fixed in fa28959e — the note now says the fiber coordinates become selectable once attached, and points at `split_by`.

File: dascore/proc/coords.py [hunk 5]
```diff
@@ -566,6 +569,88 @@ def select(
...
+@patch_function(history=None)
+def unselect(
+    patch: PatchType, *, copy=False, relative=False, samples=False, **kwargs
+) -> PatchType:
+    """
+    Return the patch outside a selection.
+
+    The complement of [`Patch.select`](`dascore.Patch.select`): it takes
+    the same selectors and removes exactly the samples that selection
+    would have kept.
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/proc/coords.py]: Fixed in fa28959e. The opening sentence now says it removes the samples the selection would have kept, exact for one coordinate, and refers to the note for several; the tutorial says the same.

File: dascore/proc/coords.py [hunk 6]
```diff
@@ -566,6 +569,88 @@ def select(
...
+    >>> trimmed = patch.unselect(distance=(..., 10), samples=True)
+
+    Notes
+    -----
+    - Removing a range from the middle of a coordinate leaves a hole in
+      it, so the result is no longer evenly sampled and the coordinate
+      becomes a monotonic array. That is exactly what
+      [`Spool.unselect`](`dascore.core.spool.Spool.unselect`) refuses
+      coordinates for: at spool level the complement of a range is a hole
+      in every patch rather than a choice between patches.
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/proc/coords.py]: Fixed in fa28959e, in both places, along with `BaseSpool.unselect` — which the internal review had also flagged, since it is the composed user-facing text and still said coordinates were refused outright.

File: dascore/proc/inventory.py [hunk 7]
```diff
@@ -770,6 +784,345 @@ def _get_coord_values(inventory, path, name, distances):
...
+        # patch states a negative one, and counting samples with it would
+        # give none at all.
+        step = abs(step)
+        key = (id(context), axis, low, high, step)
+        if (placed := cache.get(key)) is None:
+            grid = np.arange(round((high - low) / step) + 1) * step + low
+            placed = (grid, context.acquisition.channel_to_distance(grid, axis=axis))
+            cache[key] = placed
+        grid, distances = placed
+        yield PlacedRow(context, low, high, grid, distances, None)
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/proc/inventory.py:1010]: Investigated and did not reproduce, so this is a test rather than a fix (8ea1f935). Over six pathological float grids — `1.0 + i*0.1`, `0.1 + i*0.25`, `-1234.5678 + i*0.037`, `i/3`, and two others — the channels the index selects are exactly the channels enrichment projects the value onto, same count and same positions every time.

There *is* a difference, but it is not this one and it is not new: the loaded values differ from the originals in the last ulp, because a trimmed `CoordRange` regenerates from the piece's own start. A plain `patch.select(distance=(a, b))` with no inventory anywhere shows it too (`10.299999999999967` against `10.3`, max deviation 7.5e-14), so it is how any trim behaves rather than anything channel selection introduces.

`TestChannelReviewFindings::test_the_index_selects_what_enrichment_projects` now pins the property, compared by count and position rather than by value for that reason.
